runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsResponseInterceptor.kt

```diff
@@ -31,60 +32,85 @@ internal val CHECKSUM_HEADER_VALIDATION_PRIORITY_LIST: List<String> = listOf(
-            return context.protocolResponse
-        }
+        val logger = coroutineContext.logger<FlexibleChecksumsResponseInterceptor>()
 
-        val logger = coroutineContext.logger<FlexibleChecksumsResponseInterceptor<I>>()
+        val configuredToVerifyChecksum = responseValidationRequired || responseChecksumValidation == HttpChecksumConfigOption.WHEN_SUPPORTED
+        if (!configuredToVerifyChecksum) return context.protocolResponse
 
         val checksumHeader = CHECKSUM_HEADER_VALIDATION_PRIORITY_LIST
             .firstOrNull { context.protocolResponse.headers.contains(it) } ?: run {
-            logger.warn { "User requested checksum validation, but the response headers did not contain any valid checksums" }
+            logger.warn { "Checksum validation was requested but the response headers didn't contain a valid checksum." }
             return context.protocolResponse
         }
 
-        // let the user know which checksum will be validated
-        logger.debug { "Validating checksum from $checksumHeader" }
-        context.executionContext[ChecksumHeaderValidated] = checksumHeader
+        val serviceChecksumValue = context.protocolResponse.headers[checksumHeader]!!
+        if (ignoreChecksum(serviceChecksumValue)) {
+            logger.warn { "Checksum detected but validation was skipped." }
+            return context.protocolResponse
+        }
 
-        val checksumAlgorithm = checksumHeader.removePrefix("x-amz-checksum-").toHashFunction() ?: throw ClientException("could not parse checksum algorithm from header $checksumHeader")
+        context.executionContext[ChecksumHeaderValidated] = checksumHeader
 
-        // Wrap the response body in a hashing body
-        return context.protocolResponse.copy(
-            body = context.protocolResponse.body
-                .toHashingBody(checksumAlgorithm, context.protocolResponse.body.contentLength)
-                .toChecksumValidatingBody(context.protocolResponse.headers[checksumHeader]!!),
-        )
+        val checksumAlgorithm = checksumHeader
+            .removePrefix("x-amz-checksum-")
+            .toHashFunctionOrThrow()
+
+        when (val bodyType = context.protocolResponse.body) {
+            is HttpBody.Bytes -> {
+                logger.debug { "Validating checksum before deserialization from $checksumHeader" }
+
+                checksumAlgorithm.update(
+                    context.protocolResponse.body.readAll() ?: byteArrayOf(),
+                )
+                val sdkChecksumValue = checksumAlgorithm.digest().encodeBase64String()
+
+                validateAndThrow(
+                    serviceChecksumValue,
+                    sdkChecksumValue,
+                )
+
+                return context.protocolResponse
+            }
+            is HttpBody.SourceContent, is HttpBody.ChannelContent -> {
+                logger.debug { "Validating checksum during deserialization from $checksumHeader" }
```

Comment:
misleading: It's validated **after** deserialization, after the user is done reading the entire body.